### DISCUSSION TITLE
Cannot determine a partition for slot xyz

### DISCUSSION THREAD

**Discussion by priyankaexp** [Grokzen/docker-redis-cluster#141] [Q&A] [answered]
I am using Lettuce RedisClusterClient to connect to the docker container.  Any query that I put in via lettuce client, I get the error `Cannot determine a partition for slot xxxx`. 

I can do redis-cli though and `cluster slots` looks ok. And the commands as well 

```
127.0.0.1:7001> cluster slots
1) 1) (integer) 5461
   2) (integer) 10922
   3) 1) "172.17.0.3"
      2) (integer) 7001
      3) "6af66914dab9106b02e5b9966f4a1c0ffd2124af"
   4) 1) "172.17.0.3"
      2) (integer) 7004
      3) "9d6643629971a546a536983e6dd7fda7e1453e74"
2) 1) (integer) 0
   2) (integer) 5460
   3) 1) "172.17.0.3"
      2) (integer) 7000
      3) "3de61f61a64b8c49fc2f1ed8bec543760efbdb25"
   4) 1) "172.17.0.3"
      2) (integer) 7003
      3) "4d3cdbf53d901ba5a3b8acec1d448be12f221d7c"
3) 1) (integer) 10923
   2) (integer) 16383
   3) 1) "172.17.0.3"
      2) (integer) 7002
      3) "b774dd3ee32b136fda7dda35b1dd5def87314f71"
   4) 1) "172.17.0.3"
      2) (integer) 7005
      3) "76b641d59f7a985c86526e92ae5f512adfa18f6f"
```

 lettuce client code 

 ```
val redisUriBuilder: RedisURI.Builder = RedisURI.builder().withHost(config.host).withPort(7000)
           val client: RedisClusterClient = RedisClusterClient.create(redisUriBuilder.build())
           val conn: StatefulRedisClusterConnection<String, String>  = client.connect(StringCodec.UTF8)
           conn.coroutines().set("key", "value") 
```

I can connect to elastic cache cluster with this code though. Any idea what I may be missing here?

**Comment by Grokzen** [comment, marked as answer]
So in general this is probably a issue with the client code, so i should say that you should ask in the lettuce domain what the issue really is.

But from experience i can most certainly say that your issue is one of the most common issues when running this container.

When running redis in cluter mode, each server node inteenrally will keep this cluster slots mapping that tells each redis server how each server reaches all other nodes from their pov. not from the clients pov. So the issues occurs when in this docker container we run all 6 redis servers on the same machine because it usually wants to track each other node to 127.0.0.1 as that is the closest route to it's neighbors. The problem with that is when a client connects and asks for the current cluster state, it will get back "oh hey client, this cluster nodes is on 127.0.0.1:7000 and not on the external ip address ourside the docker container.

This do not happen when you run on elastic cache as each redis server reaches each other node over the internet or within your vpc inside your cloud. In those cases clients and server nodes works with same IP.

One simple solution might be just to run your docker container with host-network mode and ensure that each redis node uses 127.0.0.1 as the IP it tracks for each node. The downside for this is that you removes the possibility for reaching the redis servers from any other machine then localhost. This is by design with the purpose and use-case for this docker container.

TLDR is to ensure the IP internally and externally  is reachable from your clients.
